File: src/com/kingmang/lazurite/parser/optimization/SummaryOptimization.java
package com.kingmang.lazurite.parser.optimization;

import com.kingmang.lazurite.parser.ast.Node;

public final class SummaryOptimization implements Optimizable {

    private final Optimizable[] optimizations;

    public SummaryOptimization(Optimizable[] optimizations) {
        this.optimizations = optimizations;
    }

    @Override
    public Node optimize(Node node) {
        for (Optimizable optimization : optimizations) {
            node = optimization.optimize(node);
        }
        return node;
    }

    @Override
    public int optimizationsCount() {
        int count = 0;
        for (Optimizable optimization : optimizations) {
            count += optimization.optimizationsCount();
        }
        return count;
    }

    @Override
    public String summaryInfo() {
        final StringBuilder sb = new StringBuilder();
        for (Optimizable optimization : optimizations) {
            sb.append(optimization.summaryInfo());
        }
        return sb.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/optimization/SummaryOptimization.kt
package com.kingmang.lazurite.optimization

import com.kingmang.lazurite.parser.ast.Node


class SummaryOptimization(private val optimizations: Array<Optimizable>) : Optimizable {
    override fun optimize(node: Node?): Node? {
        var node = node
        for (optimization in optimizations) {
            node = optimization.optimize(node)
        }
        return node
    }

    override fun optimizationsCount(): Int {
        var count = 0
        for (optimization in optimizations) {
            count += optimization.optimizationsCount()
        }
        return count
    }

    override fun summaryInfo(): String {
        val sb = StringBuilder()
        for (optimization in optimizations) {
            sb.append(optimization.summaryInfo())
        }
        return sb.toString()
    }
}
